base/bucket_gocb.go

```diff
@@ -1668,11 +1673,16 @@ func (bucket *CouchbaseBucketGoCB) WriteUpdate(k string, exp uint32, callback sg
 		if pkgerrors.Cause(err) == gocb.ErrKeyExists {
 			// retry on cas failure
 		} else if isRecoverableGoCBError(err) {
-			// retry on recoverable failure
+			// retry on recoverable failure, up to MaxNumRetries
+			if retryAttempts >= bucket.spec.MaxNumRetries {
+				return 0, err
+			}
 		} else {
 			// err will be nil if successful
 			return uint64(casGoCB), err
 		}
+
+		retryAttempts++
```

Comment:
Are there scenarios that warrant an exponential backoff here? Tmpfail, for example? Otherwise, will we rattle through the retries as fast as the for loop can go?

Reply:
Discussed this - since the next time through the update loop is going to include a GET for the doc, we can rely on the backoff handling in place there to deal with Tmpfail type scenarios.